internal/discharger/login_sublocation_test.go

```diff
@@ -0,0 +1,86 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package discharger_test
+
+import (
+	"net/http"
+	"path/filepath"
+	"testing"
+
+	qt "github.com/frankban/quicktest"
+	"github.com/frankban/quicktest/qtsuite"
+
+	"github.com/canonical/candid/idp"
+	"github.com/canonical/candid/idp/static"
+	"github.com/canonical/candid/internal/candidtest"
+	"github.com/canonical/candid/internal/discharger"
+	"github.com/canonical/candid/internal/identity"
+)
+
+const sublocationPath = "/sublocation"
+
+func TestLoginWithSublocation(t *testing.T) {
+	qtsuite.Run(qt.New(t), &loginCookieSuite{})
+}
+
+type loginCookieSuite struct {
+	params *identity.ServerParams
+}
+
+func (s *loginCookieSuite) Init(c *qt.C) {
+	store := candidtest.NewStore()
+	sp := store.ServerParams()
+	sp.IdentityProviders = []idp.IdentityProvider{
+		static.NewIdentityProvider(static.Params{
+			Name:   "test",
+			Domain: "test",
+			Icon:   "/static/static1.bmp",
+		}),
+	}
+	s.params = &sp
+}
+
+func (s *loginCookieSuite) TestLoginCookiePathContainsServerSublocation(c *qt.C) {
+	s.params.UseLocationForCookiePaths = true
+	srv := candidtest.NewServerWithSublocation(c, *s.params, map[string]identity.NewAPIHandlerFunc{
+		"discharger": discharger.NewAPIHandler,
+	}, sublocationPath)
+
+	req, err := http.NewRequest("GET", sublocationPath+"/login", nil)
+	c.Assert(err, qt.IsNil)
+	req.Header.Set("Accept", "application/json")
+	resp := srv.Do(c, req)
+	defer resp.Body.Close()
+	c.Assert(resp.StatusCode, qt.Equals, http.StatusOK)
+
+	cookies := resp.Cookies()
+	c.Assert(len(cookies) > 0, qt.IsTrue)
+
+	for _, cookie := range cookies {
+		dir := filepath.Dir(cookie.Path)
+		c.Assert(dir, qt.Equals, sublocationPath)
+	}
+}
+
+func (s *loginCookieSuite) TestLoginCookiePathNotContainsServerSublocation(c *qt.C) {
+	s.params.UseLocationForCookiePaths = false
+	srv := candidtest.NewServerWithSublocation(c, *s.params, map[string]identity.NewAPIHandlerFunc{
+		"discharger": discharger.NewAPIHandler,
+	}, sublocationPath)
+
+	req, err := http.NewRequest("GET", sublocationPath+"/login", nil)
+	c.Assert(err, qt.IsNil)
+	req.Header.Set("Accept", "application/json")
+	resp := srv.Do(c, req)
+	defer resp.Body.Close()
+	c.Assert(resp.StatusCode, qt.Equals, http.StatusOK)
+
+	cookies := resp.Cookies()
+	c.Assert(len(cookies) > 0, qt.IsTrue)
+
+	for _, cookie := range cookies {
+		dir := filepath.Dir(cookie.Path)
+		c.Assert(dir != sublocationPath, qt.IsTrue)
+	}
+}
```

Comment:
Cool tests, but do you really need a suite for them or would table testing be sufficient?
```go
var cookiePathTests = []struct{
	about string
	useLocation bool
}{{
	about: "location in the cookie",
	useLocation: true,
}, {
	about: "location NOT in the cookie",
}}

func TestLoginCookiePath(t *testing.T) {
	c := qt.New(t)
	for _, test := range cookiePathTests {
		c.Run(test.about, func(c *qt.C) {
			// Set up the store and the server.
			store := candidtest.NewStore()
			p := store.ServerParams()
			p.IdentityProviders = []idp.IdentityProvider{
				static.NewIdentityProvider(static.Params{
					Name:   "test",
					Domain: "test",
					Icon:   "/static/static1.bmp",
				}),
			}
			p.UseLocationForCookiePaths = test.useLocation
			srv := candidtest.NewServerWithSublocation(c, p, map[string]identity.NewAPIHandlerFunc{
				"discharger": discharger.NewAPIHandler,
			}, sublocationPath)

			// Make the request.
			req, err := http.NewRequest("GET", sublocationPath+"/login", nil)
			c.Assert(err, qt.IsNil)
			req.Header.Set("Accept", "application/json")
			resp := srv.Do(c, req)
			defer resp.Body.Close()

			// Check the response.
			c.Assert(resp.StatusCode, qt.Equals, http.StatusOK)
			cookies := resp.Cookies()
			c.Assert(cookies, qt.Not(qt.HasLen), 0)
			for _, cookie := range cookies {
				dir := filepath.Dir(cookie.Path)
				if test.useLocation {
					c.Assert(dir, qt.Equals, sublocationPath)
				} else {
					c.Assert(dir, qt.Not(qt.Equals), sublocationPath)
				}
			}
		})
	}
```
Take it or leave it, your current implementation is totally fine.

Reply:
I like your solution much better :) I did like this due to a copy-paste-edit from login_test